Add a dedicated concept page for Pod Conditions (#55316)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-condition.md

---
title: "Pod Conditions"
content_type: concept
weight: 35
---

> 

Reviewer: ```suggestion

In Kubernetes, many objects have _conditions_. Conditions are markers for some aspect of the actual state of the thing
the object represents.

Pods have conditions, and Kubernetes Pod conditions are an important aspect of how controllers
(and people doing troubleshooting)
can understand the health of a Pod.

```

A Pod's [phase](/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase) provides a high-level
summary of where the Pod is in its lifecycle, but a single value cannot capture the full
picture. For example, a Pod may be in the `Running` phase but not yet ready to serve traffic.
_Pod conditions_ complement the phase by tracking multiple aspects of the Pod's state
independently, such as whether it has been scheduled, whether its containers are ready,
or whether a resize is in progress.

> 

Reviewer: At this point I would also add one sentence that hyperlinks to the concept of a taint.

## Structure of a Pod condition

A Pod's status includes an array of
[PodConditions](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#podcondition-v1-core)
that indicate whether the Pod has passed certain checkpoints.

Each element of the PodCondition array has the following fields:

> Field name           | Description

Reviewer: nit: we like tables to have captions

No need to add that in this PR though.

:--------------------|:-----------
`type`               | Name of this Pod condition.
`status`             | Indicates whether that condition is applicable, with possible values "`True`", "`False`", or "`Unknown`".
`lastProbeTime`      | Timestamp of when the Pod condition was last probed.
`lastTransitionTime` | Timestamp for when the Pod last transitioned from one status to another.
`reason`             | Machine-readable, UpperCamelCase text indicating the reason for the condition's last transition.
`message`            | Human-readable message indicating details about the last status transition.
`observedGeneration` | The `.metadata.generation` of the Pod at the time the condition was recorded.

> {{< note >}}

Reviewer: Better: put the observed generation thing on the Pod lifecycle page; have this new page hyperlink there.

Author: The [Pods page](https://kubernetes.io/docs/concepts/workloads/pods/#pod-generation) already has a Pod generation section covering this, so I linked to that instead

The kubelet updates the Pod's status conditions to indicate the state of a resize request:

- `type: PodResizePending`: The kubelet cannot immediately grant the request. The `message` field provides an explanation of why.
  - `reason: Infeasible`: The requested resize is impossible on the current node (for example, requesting more resources than the node has).
  - `reason: Deferred`: The requested resize is currently not possible, but might become feasible later (for example if another pod is removed). The kubelet will retry the resize.
- `type: PodResizeInProgress`: The kubelet has accepted the resize and allocated resources, but the changes are still being applied. This is usually brief but might take longer depending on the resource type and runtime behavior. Any errors during actuation are reported in the `message` field (along with `reason: Error`).

If the requested resize is _Deferred_, the kubelet will periodically re-attempt the resize, for example when another pod is removed or scaled down.

For more details on Pod resize, see [Resize CPU and Memory Resources assigned to Containers](/docs/tasks/configure-pod-container/resize-container-resources/).

> ## Pod readiness {#pod-readiness}

Reviewer: ```suggestion
## Enhanced Pod readiness
```
(aside: before GA, this was known as Pod readiness++)

> Your application can inject extra feedback or signals into PodStatus: _Pod readiness_. To use this, set `readinessGates` in the Pod's `spec` to specify a list of additional conditions that the kubelet evaluates for Pod readiness.

Reviewer: Optional further improvement:
```suggestion
Your application can inject extra feedback or signals into the Pod's `.status`;
this is known as _enhanced Pod readiness_.
To use this, set `readinessGates` in the Pod's `spec` to specify a list of additional conditions that the kubelet evaluates for Pod readiness.
You then implement, or install, a controller that manages these custom conditions, and the kubelet
uses that as an extra input to decide if the Pod is
ready.
```

Readiness gates are determined by the current state of `status.condition` fields for the Pod. If Kubernetes cannot find such a condition in the `status.conditions` field of a Pod, the status of the condition is defaulted to "`False`".

```yaml
kind: Pod
...
spec:
  readinessGates:
    - conditionType: "www.example.com/feature-1"
status:
  conditions:
    - type: Ready                              # a built-in PodCondition
      status: "False"
      lastProbeTime: null
      lastTransitionTime: 2018-01-01T00:00:00Z
    - type: "www.example.com/feature-1"        # an extra PodCondition
      status: "False"
      lastProbeTime: null
      lastTransitionTime: 2018-01-01T00:00:00Z
  containerStatuses:
    - containerID: docker://abcd...
      ready: true
...
```

The Pod conditions you add must have names that meet the Kubernetes [label key format](/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).

### Status for Pod readiness

> The `kubectl patch` command does not support patching object status. To set these `status.conditions` for the Pod, applications and {{< glossary_tooltip term_id="operator-pattern" text="operators">}} should use the `PATCH` action. You can use a [Kubernetes client library](/docs/reference/using-api/client-libraries/) to write code that sets custom Pod conditions for Pod readiness.

Reviewer: nit: this may be stale info; I think you perhaps can patch it now, via the status subresource.

Author: I confirmed locally that `kubectl patch --subresource=status` works for setting custom conditions.
I updated this section.

For a Pod that uses custom conditions, that Pod is evaluated to be ready **only** when both the following statements apply:

- All containers in the Pod are ready.
- All conditions specified in `readinessGates` are `True`.

> When a Pod's containers are Ready but at least one custom condition is missing or `False`, the kubelet sets the Pod's `Ready` [condition](#lifecycle-pod-conditions) to `ContainersReady`.

Reviewer: This doesn't sound correct. Is it actually set to False with `reason` ContainersReady?

Author: I also thought this was unclear. I checked the actual behavior locally. When readiness gates are not satisfied, the Ready condition is set to status: False with reason: ReadinessGatesNotReady.
